ompi/op/op.h

```diff
@@ -505,37 +516,7 @@ static inline void ompi_op_reduce(ompi_op_t * op, const void *source,
                                   ompi_datatype_t * dtype)
 {
     MPI_Fint f_dtype, f_count;
-    int count = full_count;
-
-    /*
-     * If the full_count is > INT_MAX then we need to call the reduction op
-     * in iterations of counts <= INT_MAX since it has an `int *len`
-     * parameter.
-     *
-     * Note: When we add BigCount support then we can distinguish between
-     * a reduction operation with `int *len` and `MPI_Count *len`. At which
-     * point we can avoid this loop.
-     */
-    if( OPAL_UNLIKELY(full_count > INT_MAX) ) {
-        size_t done_count = 0, shift;
-        int iter_count;
-        ptrdiff_t ext, lb;
-
-        ompi_datatype_get_extent(dtype, &lb, &ext);
-
-        while(done_count < full_count) {
-            if(done_count + INT_MAX > full_count) {
-                iter_count = full_count - done_count;
-            } else {
-                iter_count = INT_MAX;
-            }
-            shift = done_count * ext;
-            // Recurse one level in iterations of 'int'
-            ompi_op_reduce(op, (const char*)source + shift, (char*)target + shift, iter_count, dtype);
-            done_count += iter_count;
-        }
-        return;
-    }
+    int count = (int)full_count;
```

Comment:
I think this is still flaky. If the compiler sees this assignment it can assume (because of UB) that it does not cause an overflow (and thus `full_count <= INT_MAX`) so this jeopardizes the check below. I'm not even sure we still need `count` here. When we do need to pass an `int` instead of `size_t` we should have a local variable `count` and the assignment there, only if we know for sure that we don't overflow.